x-pack/plugins/encrypted_saved_objects/server/crypto/encrypted_saved_objects_service.ts

```diff
@@ -528,11 +533,22 @@ export class EncryptedSavedObjectsService {
         );
       }
       if (!encryptionAADs.length) {
-        encryptionAADs.push(this.getAAD(typeDefinition, descriptor, attributes));
-        if (params?.convertToMultiNamespaceType && descriptor.namespace) {
-          // This is happening during a migration; create an alternate AAD for decrypting the object attributes by stripping out the namespace from the descriptor.
-          const { namespace, ...alternateDescriptor } = descriptor;
-          encryptionAADs.push(this.getAAD(typeDefinition, alternateDescriptor, attributes));
+        if (params?.isTypeBeingConverted) {
+          // The object is either pending conversion to a multi-namespace type, or it was just converted. We may need to attempt to decrypt
+          // it with several different descriptors depending upon how the migrations are structured, and whether this is a full index
+          // migration or a single document migration. Note that the originId is set either when the document is converted _or_ when it is
+          // imported with "createNewCopies: false", so we have to try with and without it.
+          const ids = params.originId ? [params.originId, descriptor.id] : [descriptor.id];
+          for (const id of ids) {
+            const decryptDescriptor = { ...descriptor, id };
```

Comment:
> **question:** in case `params?.isTypeBeingConverted === true` and `originId` is present, do we assume that AAD with `originId` has a higher chance to be the correct one so that we put it first in the array?

Yes, that's the idea!